### ISSUE TITLE
access training set in order

### ISSUE THREAD

**Issue by jeff-regier** [YosefLab/scVI#141]
During training, we're accessing the training set out of order (i.e., not as contiguous blocks). For the retina dataset, the follow line of code takes 70% of the total runtime:

https://github.com/YosefLab/scVI/blob/0683bf5b7b947873f5b3b2c84243400452a62cec/scvi/dataset/dataset.py#L55

We can greatly reduce the data loading runtime by shuffling the training set once per epoch, and then accessing it as contiguous blocks. e.g.,
```
torch.from_numpy(self.X[index:(index + batch_size)])
```

- `torch.from_numpy` seems to be a lot faster than `torch.FloatTensor` ... presumably only the latter makes a copy
- But for `torch.from_numpy` to work, `X` needs to be stored as `np.float32`
- We also need `X` to be contiguous for this to work, which we can get with `np.ascontiguousarray(X, dtype=np.float32)`
- We have to store the training set first in `X` for this to work -- and then just shuffle for training set part of `X`
- We probably don't need `GeneExpressionDataset` to support sparse arrays any more: we always seem to take a subset of BRAIN_LARGE anyway. 720 genes * 4 bytes * 1.3 million cells = ~4 GB.
- `BrainLargeDataset` would still load the dataset as a sparse array. But it would convert it to dense after subsetting the 720 top genes.
